core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,666 @@
+              arg match {
+                case NormalExpression.Struct(_, args) =>
+                  processArgs(args, acc)
+                case _ =>
+                  NotProvable
+              }
+            } 
+
+          case Some(idx) =>
+            // we don't check if idx < 0, because if we compiled, it can't be
+            val result = { (arg: NormalExpression, acc: PatternEnv) =>
+              arg match {
+                case NormalExpression.Struct(enumId, args) =>
+                  if (enumId == idx) processArgs(args, acc)
+                  else NoMatch
+                case _ =>
+                  NotProvable
+              }
+            }
+            result
+        }
+  }
+
+  def findMatch(m: NormalExpression.Match) =
+    m.branches.collectFirst(Function.unlift( { case (pat, result) =>
+      maybeBind(pat).apply(m.arg, Map()) match {
+        case Matches(env) => Some(Some((pat, env, result)))
+        case NotProvable => Some(None)
+        case NoMatch => None
+      }
+    })).get // Totallity of matches should ensure this will always find something unless something has gone terribly wrong
+
+  def solveMatch(env: PatternEnv, result: NormalExpression) =
+    ((env.size - 1) to 0 by -1).toList.map(env.get(_).get) // If this exceptions then somehow we didn't get enough names in the env
+      .foldLeft(result) { case (ne, arg) => NormalExpression.App(ne, arg) }
+ 
+  def normalOrderReduction(expr: NormalExpression): NormalExpression = {
+    import NormalExpression._
+    val nextExpr = expr match {
+      case App(Lambda(nextExpr), arg) => {
+        applyLambdaSubstituion(nextExpr, Some(arg), 0)
+      }
+      case m@Match(struct@Struct(enum, args), branches) =>
+        findMatch(m) match {
+          case None => m
+          case Some((pat, env, result)) => solveMatch(env, result)
+        }
+      case Recursion(Lambda(innerExpr)) if(innerExpr.maxLambdaVar.map(_ < 0).getOrElse(true)) => {
+        applyLambdaSubstituion(innerExpr, None, 0)
+      }
+      case Lambda(App(innerExpr, LambdaVar(0))) => {
+        innerExpr
+      }
+      case _ => expr
+    }
+    val res = nextExpr match {
+      case al @ App(Lambda(_), _) => normalOrderReduction(al)
+      case App(fn, arg) =>
+        normalOrderReduction(fn) match {
+          case l @ Lambda(_) => normalOrderReduction(App(l, arg))
+          case nfn @ _       => App(nfn, normalOrderReduction(arg))
+        }
+      case extVar @ ExternalVar(_, _) => extVar
+      case Match(arg, branches) =>
+        val nextMatch = Match(normalOrderReduction(arg), branches)
+        findMatch(nextMatch) match {
+          case None => nextMatch
+          case Some(_) => normalOrderReduction(nextMatch)
+        }
+      case lv @ LambdaVar(_)  => lv
+      case Lambda(expr)       =>
+        normalOrderReduction(expr) match {
+          case a@App(innerExpr, LambdaVar(0)) => normalOrderReduction(Lambda(a))
+          case na @ _ => Lambda(na)
```

Comment:
it seems to me this should be safe:
```
val exp1 = normalOrderReduction(exp)
if (exp1 != exp) normalOrderReduction(Lambda(exp1))
else {
  Lambda(exp)
}
```
Not clear to me why that doesn't work.

Reply:
I think it does? Let me get a few more tests then I can give it a try